### PR TITLE
expression, util: DecodeDecimal return real precision and frac

### PR DESCRIPTION
**Pull request by winkyao**
Because we use passing params to set precision and frac in `codec.EncodeDecimal`, so we need to get the real decimal precision and frac from `codec.DecodeDecimal`, to solve the admin check table failed:

```
mysql root@127.0.0.1:test> CREATE TABLE test (  a decimal, PRIMARY KEY (a));
Query OK, 0 rows affected
Time: 0.102s
mysql root@127.0.0.1:test> insert into test set a=10;
Query OK, 1 row affected
Time: 0.124s
mysql root@127.0.0.1:test> select * from test;
+----+
| a  |
+----+
| 10 |
+----+
1 row in set
Time: 0.008s
mysql root@127.0.0.1:test> admin check table test;
(1105, 'test err:[admin:1]index:<nil> != record:&{1 [{8 0 0 2 0 [] 0xc421406120}]}')
```

### CODE REVIEW

**Comment by shenli** [comment]
@winkyao Please resolve the conflicts.

File: util/codec/decimal_test.go [hunk 1]
```diff
@@ -48,7 +48,7 @@ func (s *testDecimalSuite) TestDecimalCodec(c *C) {
 		v := types.NewDecFromFloatForTest(input.Input)
 		datum := types.NewDatum(v)
 		b := EncodeDecimal([]byte{}, datum.GetMysqlDecimal(), datum.Length(), datum.Frac())
-		_, d, err := DecodeDecimal(b)
```
Comment (XuHuaiyu) [inline util/codec/decimal_test.go]: add test for precision and frac here.

**Comment by jackysp** [comment]
LGTM

File: util/codec/decimal_test.go (hunk 1 above)
Comment (zhexuany) [reply util/codec/decimal_test.go]: If `DecodeDecimal` always retuned legitimate `precision` or `frace`, then there is no need to add such tests. Alternative way to test could be check the error which is retuned by `DecodeDecimal`.

**Comment by winkyao** [comment]
@XuHuaiyu PTAL

File: util/codec/decimal_test.go [hunk 2]
```diff
@@ -47,9 +47,18 @@ func (s *testDecimalSuite) TestDecimalCodec(c *C) {
...
+
 		b, err := EncodeDecimal([]byte{}, datum.GetMysqlDecimal(), datum.Length(), datum.Frac())
 		c.Assert(err, IsNil)
-		_, d, err := DecodeDecimal(b)
+		_, d, prec, frac, err := DecodeDecimal(b)
+		if datum.Length() != 0 {
+			c.Assert(prec, Equals, datum.Length())
+			c.Assert(frac, Equals, datum.Frac())
+		} else {
+			prec1, frac1 := datum.GetMysqlDecimal().PrecisionAndFrac()
```
Comment (XuHuaiyu) [inline util/codec/decimal_test.go]: s/datum.GetMysqlDecimal().PrecisionAndFrac()/ v.PrecisionAndFrac()

File: util/codec/decimal_test.go [hunk 3]
```diff
@@ -47,9 +47,18 @@ func (s *testDecimalSuite) TestDecimalCodec(c *C) {
 	for _, input := range inputs {
 		v := types.NewDecFromFloatForTest(input.Input)
 		datum := types.NewDatum(v)
+
 		b, err := EncodeDecimal([]byte{}, datum.GetMysqlDecimal(), datum.Length(), datum.Frac())
 		c.Assert(err, IsNil)
-		_, d, err := DecodeDecimal(b)
+		_, d, prec, frac, err := DecodeDecimal(b)
+		if datum.Length() != 0 {
```
Comment (XuHuaiyu) [inline util/codec/decimal_test.go]: always be false?

File: util/codec/decimal_test.go (hunk 2 above)
Comment (winkyao) [reply util/codec/decimal_test.go]: datum.GetMysqlDecimal().PrecisionAndFrac() is ok.

**Pull request merged**